### PR TITLE
Backend: order controller - revise createOrder, add getOrdersByUser and getOrdersByShop

### PR DESCRIPTION
**Pull request by VvnChengg**
1. Revise createOrder: allows clients to create orders along with associated order items in a single request
2. Add getOrdersByUser: extract from getOrders
3. Add getOrdersByShop: extract from getOrders
4. OrderStatus add "cancelled"

### CODE REVIEW

File: backend/src/routes/order.ts [hunk 1]
```diff
@@ -5,13 +5,19 @@ import {
...
 } from '../controllers/order';
 
 const router = express.Router();
 
 // GET /api/order/
 router.get('/', getOrders);
+// GET /api/order/user/:user_id
+router.get('/user/:user_id', getOrdersByUser);
+// GET /api/order/shop/:shop_id
+router.get('/shop/:shop_id', getOrdersByShop);
```
Comment (slimon110) [inline backend/src/routes/order.ts:20]: 我記得這兩個 api 昨天討論的結果是歸在 /api/shop & /api/user 下面欸

Comment (slimon110) [reply backend/src/routes/order.ts:20]: 但仔細想想放在 order 下面是不是更合理？

Comment (VvnChengg) [reply backend/src/routes/order.ts:20]: 是要放在api/shop和api/user底下沒錯，我這邊的router只是測試功能正常，確保你們可以用

File: backend/src/controllers/user.ts [hunk 2]
```diff
@@ -126,3 +129,17 @@ export const deleteUser = async (
         genericErrorHandler(err, res);
     }
 };
+
+export const getOrdersByUser = async (
```
Comment (slimon110) [inline backend/src/controllers/user.ts:133]: What about `getOrdersByUserId` and `getOrdersByShopId`?

File: backend/src/controllers/order_repository.ts [hunk 3]
```diff
@@ -24,28 +24,28 @@ interface IOrderReposiotry {
 }
 
 export class MongoOrderRepository implements IOrderReposiotry {
-    async findAll(): Promise<GetOrdersResponse | null> {
+    async findAll(): Promise<GetOrdersResponse> {
```
Comment (slimon110) [inline backend/src/controllers/order_repository.ts:27]: Remember to also modify the IOrderRepository

Comment (slimon110) [reply backend/src/controllers/order_repository.ts:27]: I mean the return type.

File: backend/src/controllers/order_repository.ts [hunk 4]
```diff
@@ -23,29 +23,29 @@ interface IOrderReposiotry {
...
         return OrderModel.findById(id);
     }
 
-    async findByUserId(id: string): Promise<GetOrdersResponse | null> {
+    async findByUserId(id: string): Promise<GetOrdersResponse> {
         return OrderModel.find({ user_id: id });
     }
 
-    async findByShopId(id: string): Promise<GetOrdersResponse | null> {
+    async findByShopId(id: string): Promise<GetOrdersResponse> {
```
Comment (slimon110) [inline backend/src/controllers/order_repository.ts:39]: And this?

File: backend/src/controllers/user.ts (hunk 2 above)
Comment (VvnChengg) [reply backend/src/controllers/user.ts:133]: 已改！

File: backend/src/controllers/order_repository.ts (hunk 3 above)
Comment (VvnChengg) [reply backend/src/controllers/order_repository.ts:27]: done!

**Review by slimon110 — APPROVED** [review]
LGTM.

**Pull request merged**